### PR TITLE
[BEAM-853] Force streaming execution on batch pipelines for testing.

### PR DESCRIPTION
**Pull request by amitsela**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by amitsela** [comment]
This is the first test in executing ROS for streaming - being able to execute `BoundedReadFromUnboundedSource` in a streaming pipeline.

R: @kennknowles

File: runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java [hunk 1]
```diff
@@ -65,11 +77,23 @@ public static TestSparkRunner fromOptions(PipelineOptions options) {
...
-  public <OutputT extends POutput, InputT extends PInput>
-      OutputT apply(PTransform<InputT, OutputT> transform, InputT input) {
-    return delegate.apply(transform, input);
-  };
+  public <OutputT extends POutput, InputT extends PInput> OutputT apply(
+          PTransform<InputT, OutputT> transform, InputT input) {
+    // if the pipeline is required to execute as a streaming pipeline,
+    // and the source is an adapted unbounded source (as bounded),
+    // read it as unbounded source via UnboundedReadFromBoundedSource.
+    if (isStreaming && transform instanceof BoundedReadFromUnboundedSource) {
```
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java]: Can this be done in the translator? I know it is isolated to testing, but perhaps it can be a somewhat secret flag that you enable, like `delegate.setForceStreaming(true)`?

The reason I say this is that we actually want to eliminate `PipelineRunner.apply` so that the user completely constructs their pipeline before the runner has a handle on it.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/BoundedReadFromUnboundedSource.java [hunk 2]
```diff
@@ -83,10 +83,18 @@
     this.maxReadTime = maxReadTime;
   }
 
+  /**
+   * Returns an adapted {@link BoundedSource} wrapping the underlying {@link UnboundedSource},
+   * with the specified bounds on number of records and read time.
+   */
+  public BoundedSource<ValueWithRecordId<T>> getAdaptedSource() {
+    return new UnboundedToBoundedSourceAdapter<>(source, maxNumRecords, maxReadTime);
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/BoundedReadFromUnboundedSource.java]: Build this in the constructor?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java [hunk 3]
```diff
@@ -78,6 +102,33 @@ public SparkPipelineResult run(Pipeline pipeline) {
...
+      int success = result.getAggregatorValue(PAssert.SUCCESS_COUNTER, Integer.class);
+      assertThat("Success aggregator should be greater than zero.", success, not(0));
+      // validate assertion didn't fail.
+      int failure = result.getAggregatorValue(PAssert.FAILURE_COUNTER, Integer.class);
+      assertThat("Failure aggregator should be zero.", failure, is(0));
+    }
     return result;
   }
+
+  private static class AdaptedBoundedAsUnbounded extends PTransform<PBegin, PCollection<?>> {
```
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java]: Even though the uses of this are all dynamically typed, adding a type parameter would increase the type safety and readability within this class definition, for example preventing accidentally returning `PCollection<ValueWithRecordId<T>>`. I mean something like this:

```java
class AdaptedBoundedAsUnbounded<T> extends PTransform<PBegin, PCollection<T>> {
  ...
  AdaptedBoundedAsUnbounded(BoundedReadFromUnboundedSource<T> source) { ... }
}
```

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/BoundedReadFromUnboundedSource.java [hunk 4]
```diff
@@ -83,10 +83,18 @@
     this.maxReadTime = maxReadTime;
   }
 
+  /**
+   * Returns an adapted {@link BoundedSource} wrapping the underlying {@link UnboundedSource},
+   * with the specified bounds on number of records and read time.
+   */
+  public BoundedSource<ValueWithRecordId<T>> getAdaptedSource() {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/BoundedReadFromUnboundedSource.java]: This is a new public method that is fundamental enough that I could see users coming to depend on it. I approve of it from the position that a `PTransform` subclass should feel free to expose its definitive pieces, much like `ParDo` exposes the `DoFn`.

Can you mark it `@Experimental`? We need to start increasing our discipline about being very careful not to expose public APIs that might be considered stable too early.

File: runners/spark/src/test/java/org/apache/beam/runners/spark/ForceStreamingTest.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+      }
+
+      @Override
+      public void validate() { }
+
+      @Override
+      public Coder getDefaultOutputCoder() {
+        return null;
+      }
+
```
Comment (kennknowles) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/ForceStreamingTest.java]: nit: Lots of extra whitespace and some funky newlines elsewhere. Autoformat?

File: runners/spark/src/test/java/org/apache/beam/runners/spark/ForceStreamingTest.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+ * <p>This test does not execute a pipeline.
+ */
+public class ForceStreamingTest {
+
+  @Test
+  public void test() throws IOException {
+    SparkPipelineOptions options = PipelineOptionsFactory.create().as(SparkPipelineOptions.class);
+    options.setRunner(TestSparkRunner.class);
+    // force streaming.
+    options.setStreaming(true);
```
Comment (kennknowles) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/ForceStreamingTest.java]: Related to my comment above, you could so `options.as(SparkRunnerTestOptions.class).setForceStreaming(true)` and then read that in the transform translators.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java (hunk 1 above)
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java]: I'll check it out.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java (hunk 3 above)
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java]: 👍

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/BoundedReadFromUnboundedSource.java (hunk 4 above)
Comment (amitsela) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/BoundedReadFromUnboundedSource.java]: 👍

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/BoundedReadFromUnboundedSource.java (hunk 2 above)
Comment (amitsela) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/BoundedReadFromUnboundedSource.java]: 👍

File: runners/spark/src/test/java/org/apache/beam/runners/spark/ForceStreamingTest.java (hunk 5 above)
Comment (amitsela) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/ForceStreamingTest.java]: A messy paste indeed 😉

File: runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java (hunk 1 above)
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java]: I generally agree with with going even more towards portability, but doesn't losing `PipelineRunner.apply` mean losing the runner's ability to manipulate the `Pipeline` ? Of course the user's `Pipeline` shouldn't change logically, but runners should have the freedom to replace transforms to achieve the execution of the "same" logic, no ?

Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java]: Yes, definitely. Basically the idea is to replace `PipelineRunner.apply` with a `Pipeline.replace` that takes more-or-less the same parameters and is just as easy to use. @tgroh is actively working on this, doing preparatory refactors to `TransformHierarchy`.

But since the Spark runner already pretty much does this right - doing it in the translation visitor - you have a pretty great hook to just already do it. If it is a pain to do it there, then it is fine to do it here. Porting to the features I am describing should be easy - all runners will need their `apply` to be ported anyhow.

**Comment by amitsela** [comment]
For now, I'll keep relying on `PipelineRunner.apply` (it's not even marked as `@Deprecated` yet..).
It actually shouldn't be too hard to do this in the translator, but then I won't be able to have a test that doesn't execute a pipeline, and proper streaming tests with the SDK require triggers support etc.
Once I will have proper streaming tests, this could be done, but I'm trying to avoid huge commits.

Addressed all other comments, thanks!

**Comment by amitsela** [comment]
Squashed & rebased to prep for merge and updated to overcome (unrelated) failed tests that are fixed now.
@kennknowles PTAL.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java [hunk 7]
```diff
@@ -78,6 +102,33 @@ public SparkPipelineResult run(Pipeline pipeline) {
     result.waitUntilFinish();
     assertThat(result, testPipelineOptions.getOnCreateMatcher());
     assertThat(result, testPipelineOptions.getOnSuccessMatcher());
+    // if the pipeline was executed in streaming mode, validate aggregators.
+    if (isStreaming) {
+      // validate assertion succeeded (at least once).
+      int success = result.getAggregatorValue(PAssert.SUCCESS_COUNTER, Integer.class);
```
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java]: You can get even more accuracy here: https://github.com/apache/incubator-beam/blob/master/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java#L182

Otherwise I think 2 asserts, 1 success, 0 failures will report a success.

The better way than counting in `apply` would be with (new class owned by) `PAssert.AssertionCountingPipelineVisitor`.

I hate to push the scope of your PR bigger, so at least you could mimic the `DataflowRunner` way of counting them.

**Comment by kennknowles** [comment]
LGTM if you count assertions or have a reason why it doesn't matter.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java (hunk 7 above)
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java]: For now, I don't see a reason to extend this.
I'm pretty sure `FAILURE_COUNTER` will never show as spark Accumulators don't report back if task failed (thrown Exception), but I'm asserting this anyway.
The `SUCCESS_COUNTER` assertion is important just to make sure any assertion occurred - this is because in case no data is processed (for whatever reason) tests will succeeded without actually asserting anything.
So basically I'm asserting that _any_ assertion happened.
If needed, this could be extended in the future. For now, I'll merge.

**Comment by amitsela** [comment]
I've added a `forceStreaming` option like you suggested earlier, otherwise, it messes around with batch ROS tests.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java (hunk 7 above)
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java]: What if there's more than one assertion but one of them gets skipped just like you describe?

Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java]: It would indeed be naive of me to expect any success to ingest data into the pipeline to also mean all data was successfully ingested, so I will add the same solution as `TestDataflowRunner`.
Opened [`BEAM-1173`](https://issues.apache.org/jira/browse/BEAM-1173) for a `Visitor`-based solution like you suggested.

**Comment by kennknowles** [comment]
retest this please

**Comment by kennknowles** [comment]
Jenkins failure was breakage on `master`.

**Comment by kennknowles** [comment]
LGTM & green. Merging it.
